src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs

```diff
@@ -74,119 +87,234 @@ public override IReadOnlyList<ProjectSnapshot> Projects
+            }
+        }
 
-                // Now we need to know if the changes that we applied are significant. If that's the case then 
-                // we need to notify listeners.
-                if (snapshot.HasChangesComparedTo(original))
+        public override void WorkspaceProjectAdded(Project workspaceProject)
+        {
+            if (workspaceProject == null)
+            {
+                throw new ArgumentNullException(nameof(workspaceProject));
+            }
+
+            if (!IsSupportedWorkspaceProject(workspaceProject))
+            {
+                return;
+            }
+
+            // It's possible that the host project was already initialized, and the workspace project is showing up
+            // a bit later. If that's the case then treat this as an update.
+            if (_projects.TryGetValue(workspaceProject.FilePath, out var snapshot))
+            {
+                // We also need to check the projectId here. If this is a multi-targeting project then we are only interested
+                // in a single workspace project. Just use the one that showed up first.
+                if (snapshot.WorkspaceProject == null || snapshot.WorkspaceProject.Id == workspaceProject.Id)
                 {
-                    NotifyListeners(new ProjectChangeEventArgs(snapshot, ProjectChangeKind.Changed));
+                    WorkspaceProjectChanged(workspaceProject);
                 }
+
+                return;
             }
+
+            snapshot = new DefaultProjectSnapshot(workspaceProject);
+            _projects[workspaceProject.FilePath] = snapshot;
+
+            // We expect new projects to always be dirty but the host project isn't yet set. We don't attempt to compute
+            // any state until the project is initialized.
+            //
+            // We need to notify listeners about every project add.
+            NotifyListeners(new ProjectChangeEventArgs(snapshot, ProjectChangeKind.Added));
         }
 
-        public override void ProjectRemoved(Project underlyingProject)
+        public override void WorkspaceProjectChanged(Project workspaceProject)
         {
-            if (underlyingProject == null)
+            if (workspaceProject == null)
             {
-                throw new ArgumentNullException(nameof(underlyingProject));
+                throw new ArgumentNullException(nameof(workspaceProject));
             }
-            
-            if (_projects.TryGetValue(underlyingProject.Id, out var snapshot))
+
+            if (!IsSupportedWorkspaceProject(workspaceProject))
             {
-                _projects.Remove(underlyingProject.Id);
+                return;
+            }
 
-                // We need to notify listeners about every project removal.
-                NotifyListeners(new ProjectChangeEventArgs(snapshot, ProjectChangeKind.Removed));
+            // We also need to check the projectId here. If this is a multi-targeting project then we are only interested
```

Comment:
Hmm, definitely not necessary yet BUT given that we don't have a drop down selection of what Razor TFM to use we should probably eventually be picking the TFM that matches the one in the editor.

Reply:
This is all in flux. Right now the VS contained language infrastructure doesn't support it.

Reply:
I also considered us associating multiple workspace projects with a single HostProject.

Reply:
However, all of these ideas add substantial complexity and it's not clear what they solve. If the contained language service only supports a single TFM that's an absolute blocker.